Remove getDeviceFromPtr and allocator from Type

li-roy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21964 Remove Type dispatch
* #21963 Stop using Type in Python bindings
* #21941 Remove many usages of Type
* **#21940 Remove getDeviceFromPtr and allocator from Type**
* #21923 Move UnsafeTensorFromTH and UnsafeStorageFromTH off Type
* #21320 [BC-BREAKING] Replace Type dispatch with ATenDispatch

In preparation of deleting Type. The getDeviceFromPtr logic is only used in from_blob, so I added _from_blob as a native function to dispatch the backend-based logic.

Differential Revision: [D15893330](https://our.internmc.facebook.com/intern/diff/D15893330)

Commit:
Remove getDeviceFromPtr and allocator from Type

Commit:
Update on "Remove getDeviceFromPtr and allocator from Type"

Remove getDeviceFromPtr and allocator from Type

gh-metadata: pytorch pytorch 21940 gh/li-roy/29/head

Commit:
Update on "Remove getDeviceFromPtr and allocator from Type"

Remove getDeviceFromPtr and allocator from Type

gh-metadata: pytorch pytorch 21940 gh/li-roy/29/head

Commit:
Update on "Remove getDeviceFromPtr and allocator from Type"

Remove getDeviceFromPtr and allocator from Type

gh-metadata: pytorch pytorch 21940 gh/li-roy/29/head

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -835,6 +835,11 @@
     CPU: _frac_out_cpu
     CUDA: _frac_out_cuda
 
+- func: _from_blob(void* data, int[] sizes, int[] strides, std::function<void(void*)> deleter, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
This doesn't seem like the best solution, because native_functions is supposed to basically be python and now you need to teach codegen about these crazy things like `std::function<void(void*)> deleter` just to get type-dispatch.

Can we just hardcode this function in your registration table, i.e. just make it the equivalent of type?  Because its' really just implementation detail stuff that doesn't need to be in native_functions (outside dispatch).

ezyang commented:
deferring to @gchanan for this one

gchanan commented:
based on a chat with @li-roy I believe the current thinking is to put the minimal needed functionality onto `Context` (find the device for a pointer given deviceType), similar to defaultGenerator (https://github.com/pytorch/pytorch/blob/f8f583cbae492a1cbe6ae2e5cabdfaf2431a4175/aten/src/ATen/Context.h#L61-L72).

gchanan commented:
The advantage being that we don't need to teach codegen about C++ specific things like `std::function<void(void*)>` and we don't need to hardcode things into the dispatch table.

Commit:
Update on "Remove getDeviceFromPtr and allocator from Type"

Remove getDeviceFromPtr and allocator from Type

gh-metadata: pytorch pytorch 21940 gh/li-roy/29/head

gchanan commented on aten/src/ATen/detail/CUDAHooksInterface.h:
```diff
@@ -62,6 +62,10 @@ struct CAFFE2_API CUDAHooksInterface {
   }
 
+  virtual Device getDeviceFromPtr(void* data) const {
+    AT_ERROR("Cannot get default CUDA device without ATen_cuda library. ", CUDA_HELP);
```
this error message doesn't seem correct.

gchanan reviewed (approved):
just fix the error message, I think.

Commit:
Update on "Remove getDeviceFromPtr and allocator from Type"

Remove getDeviceFromPtr and allocator from Type

gh-metadata: pytorch pytorch 21940 gh/li-roy/29/head

Commit:
Update on "Remove getDeviceFromPtr and allocator from Type"

Remove getDeviceFromPtr and allocator from Type

gh-metadata: pytorch pytorch 21940 gh/li-roy/29/head
